src/network.rs

```diff
@@ -0,0 +1,508 @@
+						},
+						Err(err) => {
+							if let Some(QueryChannel::PutRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					QueryResult::Bootstrap(result) => match result {
+						Ok(BootstrapOk { .. }) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(())).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					_ => {},
+				},
+				_ => {},
+			},
+			SwarmEvent::NewListenAddr { address, .. } => {
+				let local_peer_id = *self.swarm.local_peer_id();
+				info!(
+					"Local node is listening on {:?}",
+					address.with(Protocol::P2p(local_peer_id.into()))
+				);
+			},
+			SwarmEvent::ConnectionEstablished {
+				peer_id, endpoint, ..
+			} => {
+				if endpoint.is_dialer() {
+					if let Some(ch) = self.pending_dials.remove(&peer_id) {
+						let _ = ch.send(Ok(()));
+					}
+				}
+				// this event is of interest,
+				// pass event to output event stream
+				self.event_sender
+					.send(Event::ConnectionEstablished { peer_id, endpoint })
+					.await
+					.expect("Event receiver not to be dropped.");
+			},
+			SwarmEvent::OutgoingConnectionError { peer_id, error } => {
+				if let Some(peer_id) = peer_id {
+					if let Some(ch) = self.pending_dials.remove(&peer_id) {
+						_ = ch.send(Err(error.into()));
+					}
+				}
+			},
+			SwarmEvent::Dialing(peer_id) => info!("Dialing {}", peer_id),
+			_ => {},
+		}
+	}
+
+	async fn handle_command(&mut self, command: Command) {
+		match command {
+			Command::StartListening { addr, sender } => {
+				_ = match self.swarm.listen_on(addr) {
+					Ok(_) => sender.send(Ok(())),
+					Err(e) => sender.send(Err(e.into())),
+				}
+			},
+			Command::AddAddress {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				self.swarm
+					.behaviour_mut()
+					.kademlia
+					.add_address(&peer_id, peer_addr.clone());
+				self.pending_kad_routing.insert(peer_id, sender);
+			},
+			Command::Dial {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				// Check if peer is not already connected
+				// Dialing connected peers could happen during
+				// the bootstrap of the first peer in the network
+				if self.swarm.is_connected(&peer_id) {
+					// just skip this dial, pretend all is fine
+					_ = sender.send(Ok(())).ok();
+					return;
+				}
+
+				if let hash_map::Entry::Vacant(e) = self.pending_dials.entry(peer_id) {
+					match self
+						.swarm
+						.dial(peer_addr.with(Protocol::P2p(peer_id.into())))
+					{
+						Ok(()) => {
+							e.insert(sender);
+						},
+						Err(e) => {
+							_ = sender.send(Err(e.into()));
+						},
+					}
+				} else {
+					todo!("Implement logic for peer thats already beeing dialed.");
```

Comment:
We need to see what happens if we redial the already dialed peers...

Reply:
This part is for a rare case of dialing something that is already previously being dialed but got stuck for whatever reason. It's rare, but it can happen.